src/stream/src/executor/source/source_executor.rs

```diff
@@ -306,119 +275,101 @@ impl<S: StateStore> SourceExecutor<S> {
-                                            let reader = self
-                                                .build_stream_source_reader(Some(
-                                                    target_state.clone(),
-                                                ))
-                                                .await
-                                                .map_err(StreamExecutorError::source_error)?;
-                                            abort_tx.send(()).unwrap();
-                                            *stream_reader.lock().await = Some(reader);
-                                            self.stream_source_splits = target_state;
-                                        }
-                                    }
-                                }
+                    let barrier = barrier?;
+                    let epoch = barrier.epoch.prev;
+                    self.take_snapshot(epoch).await?;
+
+                    if let Some(Mutation::SourceChangeSplit(mapping)) = barrier.mutation.as_deref()
+                    {
+                        if let Some(target_splits) = mapping.get(&self.actor_id).cloned() {
+                            if let Some(target_state) = self.get_diff(target_splits) {
+                                log::info!(
+                                    "actor {:?} apply source split change to {:?}",
+                                    self.actor_id,
+                                    target_state
+                                );
+
+                                // Replace the source reader with a new one of the new state.
+                                let reader = self
+                                    .build_stream_source_reader(Some(target_state.clone()))
+                                    .await?;
+                                SourceReader::replace_source_chunk_reader(&mut stream, reader);
+
+                                self.stream_source_splits = target_state;
                             }
-                            self.state_cache.clear();
-                            yield Message::Barrier(barrier)
                         }
-                        _ => unreachable!(),
                     }
+                    self.state_cache.clear();
+                    yield Message::Barrier(barrier);
                 }
+
                 Either::Right(chunk_with_state) => {
-                    let chunk_with_state =
-                        chunk_with_state.map_err(StreamExecutorError::source_error)?;
-                    if chunk_with_state.split_offset_mapping.is_some() {
-                        let mapping: HashMap<String, String> =
-                            chunk_with_state.split_offset_mapping.unwrap();
+                    let StreamChunkWithState {
+                        mut chunk,
+                        split_offset_mapping,
+                    } = chunk_with_state?;
+
+                    if let Some(mapping) = split_offset_mapping {
```

Comment:
So every time we receive a chunk, we will have such many things to do... I'm thinking of how to improve this part's efficiency.

Reply:
Would it be better if we store `stream_source_splits` as a HashMap?